examples/casestudies/inference.effekt.md

```diff
@@ -0,0 +1,617 @@
+  var current = 1.0
+  try { (program(), current) }
+  with weight { (prob) =>
+    current = current * prob;
+    resume(())
+  }
+}
+```
+
+With the effect `emit` it is also possible to limit possibly infinite loops. This allows for flexible numbers of steps to be performed with any algorithm that uses the effect `emit`.
+
+```
+def onEmit[A] { handler: A => Unit } { program: => Any / Emit[A] }: Unit = {
+  try { program(); () }
+  with Emit[A] {
+    def emit(element) = { handler(element); resume(()) }
+  }
+}
+
+def limit[A](n: Int) { program: => Any / Emit[A] }: Unit / Emit[A] = {
+  var steps = n
+  try { program(); () }
+  with Emit[A] {
+    def emit(element) = {
+      do emit(element);
+      steps = steps - 1;
+      if (steps > 0) { resume(()) }
+    }
+  }
+}
+
+def collect[A,R](n: Int){ program: () => R / Emit[A] }: List[A] = {
+  var result: List[A] = Nil()
+  var steps = n
+  try { program(); () }
+  with Emit[A] {
+    def emit(element) = {
+      result = Cons(element, result);
+      steps = steps - 1;
+      if (0 < steps) { resume(()) }
+    }
+  }
+  result.reverse
+}
+```
+
+## Rejection Sampling
+The effect `Weight` is also the basis of the rejection handling algorithm
+
+```
+def handleRejection[R] { program: () => R / weight }: R / random = {
+  try { program() }
+  with weight { (prob) =>
+    if (do random() < prob) { resume(()) }
+    else { handleRejection{ program } }
+  }
+}
+```
+
+It is possible to construct the rejection sampling algorithm with the handlers of this library:
+
+```
+def rejectionSampling[R](n: Int) { program: () => R / { sample, observe } }: Unit / { random, Emit[R] } = {
+  def loop(): Unit / Emit[R] = {
+    with handleSample
+    with handleRejection
+    with handleObserve
+    do emit(program())
+    loop()
+  }
+  loop()
+}
+```
+
+## Slice Sampling
+
+```
+def sliceSamplingAlgo[R]() { program: () => R / weight } = {
+  val (result, prob) = handleSample { handleWeight { program() }}
+
+  def step(result0: R, prob0: Probability) = handleRejection {
+    val (result1, prob1) = handleWeight { program() }
+    if (prob1 < prob0) { do weight(prob1 / prob0) }
+    (result1, prob1)
+  }
+  def loop(result: R, prob: Probability): Unit / Emit[R] = {
+    do emit(result)
+    val (result1, prob1) = step(result, prob)
+    loop(result1, prob1)
+  }
+  loop(result, prob)
+}
+```
+
+## Metropolis-Hastings
+
+### Tracing
+A trace records past samples used by an algorithm. The trace also controls where the algorithm draws samples from in the next iterations.
+Constructing traces is possible by handling the effect `Sample`, not with the default handler, but one that draws new samples and records them in a trace.
+
+```
+def handleTracing[R] { program: () => R / sample }: (R, Trace) / sample = {
+  try { (program(), []) }
+  with sample { (dist) =>
+    val d = do sample(dist);
+    val (r, trace) = resume(d);
+    (r, Cons(d, trace))
+  }
+}
```

Comment:
Feel free to ignore this, but why not use mirac6le state, like in other handlers?

```suggestion
  var trace = []
  val resul = try { program() }
  with sample { (dist) =>
    val d = do sample(dist);
    trace = Cons(d, trace)
    resume(d)
  }
  (result, trace.reverse)
}
```

Reply:
thanks, typed on my phone and untested

Reply:
I am not sure what you mean by "mirac6le state"